### DISCUSSION TITLE
Thinking about using the Kotlin operator syntax

### DISCUSSION THREAD

**Discussion by progress0407** [line/kotlin-jdsl#649] [Ideas]
What do you think about using the kotlin operator syntax?

In some cases, but not all, infixes seem to lend themselves to a slightly more intuitive interpretation than prefixes.

```kotlin
vectorA.add(vectorB)
path(Book::price).gt(BigDecimal.valueOf(100))
```

If so, the above sentence could be rewritten as follows

```kotlin
vectorA + vectorB
path(Book::price) > BigDecimal.valueOf(100)
```

Unlike QueryDSL, I think this is a sugar syntax that is only possible in Kotlin

If such syntaxes are added, it will be possible to write more intuitive and clearer syntaxes.

### Before

<img width="600" alt="image" src="https://github.com/line/kotlin-jdsl/assets/66164361/429f705d-6ab2-4036-aa79-62ba48e7330e">

<img width="600" alt="image" src="https://github.com/progress0407/progress0407/assets/66164361/2da5fa89-fcfc-451e-b735-04e0970701ba">

### After

<img width="600" alt="image" src="https://github.com/progress0407/progress0407/assets/66164361/b0edfc51-57ce-4f50-9112-76a6d20ee2f5">

<img width="600" alt="image" src="https://github.com/progress0407/progress0407/assets/66164361/a21ebb7f-fb00-4db5-9d1a-fcbbd5f470f2">

### Before

<img width="400" alt="image" src="https://github.com/progress0407/progress0407/assets/66164361/63a604af-cc48-4caf-bdb2-1df1e8ebab4d">

### After

<img width="400" alt="image" src="https://github.com/progress0407/progress0407/assets/66164361/5859b98d-f8f7-4dce-88bd-7c6aaeb1a297">

### That's all working well.

<img width="600" alt="image" src="https://github.com/line/kotlin-jdsl/assets/66164361/090266a2-dac9-4727-b41b-197e7ebb2a3b">

## Item 3 (case when infix)

> That doesn't work and might not be a good idea.

<img width="600" alt="image" src="https://github.com/line/kotlin-jdsl/assets/66164361/8b3ba6ad-3881-4e3d-b33f-c17cbe989abf">

### Right now it's a dream, but

Why not further develop the domestic K-JDSL and reach 1k?

QueryDSL is a maintenance-only project with virtually no development.
- [The sad reality of QueryDSL - 1](https://stackoverflow.com/questions/55668960/whats-the-latest-status-of-querydsl)
- [The sad reality of QueryDSL - 2](https://github.com/querydsl/querydsl/issues/2170)
- [The sad reality of QueryDSL - 3](https://www.reddit.com/r/java/comments/fjpqvf/querydsl_is_slowly_getting_back_on_track/?rdt=59557)

I think Kotlin-JDSL is easier than jOOQ.

I want to let the JVM Kotlin camp know that there is KDSL.

<img width="600" alt="image" src="https://github.com/line/kotlin-jdsl/assets/66164361/c2d41dc2-48af-4cb6-95e9-4687394dafe4">

<img width="600" alt="image" src="https://github.com/line/kotlin-jdsl/assets/66164361/24b836e8-234f-4912-996c-cb7e4d8649bf">

**Comment by shouwn** [comment]
I've been thinking about the `infix` and `operator overload` since I started planning Kotlin JDSL 3.0. But the reason I haven't worked on it until now is that I can't provide a good user interface. 

## Infix

The `infix` makes our code look a little more elegant. But does the `infix` help us when writing code? I think the `infix` makes writing code more difficult because the compiler doesn't know which operand the `infix` applies to, which can be seen in the example you gave where you used parentheses.

```kotlin
whereOr(
  ( path(Employee::employeeId) ge 1L ) and ( path(Employee::employeeId) le 2L ),
  ( path(Employee::employeeId) ge 4L ) and ( path(Employee::employeeId) le 5L ),
)
```

This causes interruptions in development. Because the operators you use later affect the operators before them. You may encounter a compile error in the middle of writing code, and you have to figure out which of the operators you used to add parentheses. Also, the more parentheses you have, the harder it is to understand the code.

To help the `infix` write code, I think the operand should be defined as Algebraic Data Type (ADT) and the operators should be closed operators. However, operands in the Kotlin JDSL cannot be defined as ADTs. That's why I don't think it's a good idea to support the `infix` in the Kotlin JDSL. 

## Operator Overload

The `operator overload` inherits the problems of the `infix`, but has another problem which is that parentheses cannot be used as operators. 

```kotlin
path(Employee::employeeId) + literalLong(1L) + literalLong(2L) // employee.employeeId + 1 + 2
path(Employee::employeeId) + ( literalLong(1L) + literalLong(2L) ) // employee.employeeId + 1 + 2
```

Since we can't provide parentheses as an operator, we can't make the parentheses the user used into an actual query. If we could, we would have to add parentheses to the result of each operator. However, as more parentheses are added, the actual query that is generated becomes harder to understand, and we have to generate more and more unnecessary parenthesis objects.  

I don't think `operator overload` is worth the tradeoff.
